### ISSUE TITLE
CA1062 false positive for nullability analysis on pattern matching.

### ISSUE THREAD

**Issue by voroninp** [dotnet/roslyn-analyzers#7620]
### Analyzer

**Diagnostic ID**: [CA1062](https://learn.microsoft.com/en-us/dotnet/fundamentals/code-analysis/quality-rules/ca1062): `Validate arguments of public methods`

### Analyzer source

**SDK**: [Built-in CA analyzers in .NET 5 SDK or later](https://learn.microsoft.com/dotnet/fundamentals/productivity/code-analysis)

**Version**: [SDK 8.0.407](https://dotnet.microsoft.com/download/dotnet/8.0)

_OR_

**NuGet Package**: [Microsoft.CodeAnalysis.NetAnalyzers](https://www.nuget.org/packages/Microsoft.CodeAnalysis.NetAnalyzers)

**Version**: 9.* resolved as 9.0.0

### Steps To Reproduce

Here's the code of the method:
```
public async ValueTask<bool> TryHandleAsync(HttpContext httpContext, Exception exception, CancellationToken cancellationToken = de
{
    httpContext.NotNull();

    if (exception is not DomainError domainException)
    {
        return false;
    }

    var statusCode = GetHttpStatusCode(domainException);

    var propertyGetters = PropertyGettersByExceptionType.GetOrAdd(
        exception.GetType(),
        exceptionType => GetGettersForClientVisibleProperties(exception, exceptionType));

    var values = propertyGetters.ToDictionary(
        g => g.CamelCasedPropertyName,
        g => g.Getter(exception));

    var problemDetails = new ProblemDetails
    {
        Title = domainException.ErrorId.ErrorCode,
        Status = (int)statusCode,
        Detail = domainException.Message,
        Extensions =
        {
            [ErrorTypeKey] = domainException.ErrorId.ErrorType.FullName,
            [ErrorCodeKey] = domainException.ErrorId.ErrorCode,
            [ValuesKey] = values,
        },
    };

    problemDetails.Extensions[ValuesKey] = values;

    httpContext.Response.StatusCode = (int)statusCode;

    var problemDetailsContext = new ProblemDetailsContext {
        HttpContext = httpContext,
        ProblemDetails = problemDetails,
        Exception = exception,
    };

    return await _problemDetailsService.TryWriteAsync(problemDetailsContext);
}
```
Despite the check using pattern matching:
```
if (exception is not DomainError domainException)
{
    return false;
}
```
I get a waning from the alanylzer

### Expected behavior

Analyzer should stay silent.

### Actual behavior

Warning

![Image](https://github.com/user-attachments/assets/9a9df867-8c20-45ac-b49a-6021abefe6c8)
